src/views/virtualmachines/tree/utils/utils.tsx

```diff
@@ -219,18 +283,118 @@ export const createTreeViewData = (
       vmName,
       vmNamespace,
       treeViewDataMap,
+      clusterName,
     ),
   );
 
-  const allNamespacesTreeItem = isAdmin
-    ? createAllNamespacesTreeItem(getResourceUrl, treeViewData, treeViewDataMap, projectMap)
-    : null;
+  if (clusterName) {
+    return [
+      createClusterLevelTreeItem(
+        getResourceUrl,
+        treeViewData,
+        treeViewDataMap,
+        projectMap,
+        clusterName,
+      ),
+    ];
+  } else {
+    const allNamespacesTreeItem = isAdmin
+      ? createAllNodesTreeItem(getResourceUrl, treeViewData, treeViewDataMap, projectMap)
+      : null;
+
+    treeDataMap.value = treeViewDataMap;
+
+    const tree = allNamespacesTreeItem ? [allNamespacesTreeItem] : treeViewData;
+
+    return tree;
+  }
+};
 
-  treeDataMap.value = treeViewDataMap;
+export const createTreeViewData = (
+  getResourceUrlMultiClusterOverride: (cluster?: string) => GetResourceUrl,
+  isAdmin: boolean,
+  pathname: string,
+  foldersEnabled: boolean,
+  projectNames?: MulticlusterResource<K8sResourceCommon>[] | string[],
+  vms?: MulticlusterResource<V1VirtualMachine>[] | V1VirtualMachine[],
+  supportsMulticluster?: boolean,
+): TreeViewDataItem[] => {
+  const { currentVMTab, vmName } = getVMInfoFromPathname(pathname);
+  const treeViewDataMap: Record<string, TreeViewDataItem> = {};
+
+  // Build the project map
+  const projectMap = buildProjectMap(
+    getResourceUrlMultiClusterOverride(),
+    vms,
+    vmName,
+    currentVMTab,
+    treeViewDataMap,
+    foldersEnabled,
+  );
+
+  if (supportsMulticluster) {
+    return createMultiClusterTreeViewData(
+      getResourceUrlMultiClusterOverride,
+      projectNames as MulticlusterResource<K8sResourceCommon>[],
+      vms as MulticlusterResource<V1VirtualMachine>[],
+      isAdmin,
+      pathname,
+      treeViewDataMap,
+      projectMap,
+    );
+  }
 
-  const tree = allNamespacesTreeItem ? [allNamespacesTreeItem] : treeViewData;
+  return createClusterTreeViewData(
+    getResourceUrlMultiClusterOverride(),
+    projectNames as string[],
+    vms as V1VirtualMachine[],
+    isAdmin,
+    pathname,
+    treeViewDataMap,
+    projectMap,
+  );
+};
 
-  return tree;
+const createMultiClusterTreeViewData = (
+  getResourceUrlMultiClusterOverride: (cluster?: string) => GetResourceUrl,
+  projectNames: MulticlusterResource<K8sResourceCommon>[],
+  vms: MulticlusterResource<V1VirtualMachine>[],
+  isAdmin: boolean,
+  pathname: string,
+  treeViewDataMap: Record<string, TreeViewDataItem>,
+  projectMap: Record<string, any>,
+): TreeViewDataItem[] => {
+  const clusterTreeItems: TreeViewDataItem[] = [];
+  const multiclusterVmsRecord = groupByCluster(vms);
+  const namespaceNameByClusterRecord = groupByCluster(projectNames);
+
+  Object.keys(multiclusterVmsRecord).forEach((clusterName) => {
```

Comment:
For your list of namespaces, you might want to use projectNames, as it will give you all the namespaces, even those that do not have vms in them.